### ISSUE TITLE
Db2 cannot handle scalar subqueries in JSON_OBJECT in the presence of a MULTISET ad-hoc converter

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#12189]
This doesn't work in Db2:

```sql
select 
  json_object(
    key 'v0' value 4,
    key 'v1' value (
      select 'a'
      from SYSIBM.DUAL
    )
  )
from SYSIBM.DUAL;
```

The above produces:

> SQL Error [42601]: An unexpected token ")" was found following "SYSIBM.DUAL     ) ".  Expected tokens may include:  "&lt;interval_qualifier>".. SQLCODE=-104, SQLSTATE=42601, DRIVER=4.28.11

Casting the value to a `VARCHAR` "fixes" it

```sql
select 
  json_object(
    key 'v0' value 4,
    key 'v1' value cast((
      select 'a'
      from SYSIBM.DUAL
    ) as varchar(32672))
  )
from SYSIBM.DUAL;
```

**Comment by lukaseder** [comment]
I seem to have already fixed this, but the fix didn't work in this case:
https://stackoverflow.com/a/64350132/521799

**Comment by lukaseder** [comment]
The fix doesn't work when there's an ad-hoc converter!

**Comment by lukaseder** [comment]
The problem is that we do an `field instanceof ScalarSubquery` check, rather than calling `Tools.uncoerce(field)`, first. It's probably always wrong to just check for something being an `instanceof ScalarSubquery`

**Comment by lukaseder** [comment]
There we go.

![image](https://user-images.githubusercontent.com/734593/125637116-ed1b7df3-f490-4101-b0cd-d1738e0e3c55.png)

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.16.0 and 3.15.1 (https://github.com/jOOQ/jOOQ/issues/12190)
